Source/WebCore/css/CSSCounterStyle.cpp

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (C) 2022 Apple Inc.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "CSSCounterStyle.h"
+
+#include "CSSCounterStyleRule.h"
+#include "CSSPrimitiveValue.h"
+#include "CSSValuePair.h"
+
+namespace WebCore {
+
+bool CounterStyle::isInRange(int value) const
+{
+    if (isAutoRange()) {
+        switch (system()) {
+        case CounterStyleSystem::Cyclic:
+        case CounterStyleSystem::Numeric:
+        case CounterStyleSystem::Fixed:
+            return true;
+        case CounterStyleSystem::Alphabetic:
+        case CounterStyleSystem::Symbolic:
+            return value >= 1;
+        case CounterStyleSystem::Additive:
+            return value >= 0;
+        case CounterStyleSystem::Extends:
+            ASSERT_NOT_REACHED();
+            return true;
+        }
+    }
+
+    for (const auto& [lowerBound, higherBound] : m_ranges) {
+        if (value >= lowerBound && value <= higherBound)
+            return true;
+    }
+    return false;
+}
+
+static CounterStyle::Ranges translateRangeFromStyleRule(const StyleRuleCounterStyle& rule)
+{
+    auto ranges = rule.rangeCSSValue();
+    // auto range will return an empty Ranges
+    if (!ranges || !is<CSSValueList>(ranges))
+        return { };
+    auto& list = downcast<CSSValueList>(*ranges);
+    CounterStyle::Ranges result;
+    for (auto& rangeValue : list) {
+        ASSERT(rangeValue->isPrimitiveValue());
```

Comment:
Having to parse these from a style declaration in StyleRuleCounterStyle is awkward. I think we could have done this at parse time and just saved a CounterStyle (or a type that just contains the parsed fields, CounterStyleDescriptor or something) in StyleRuleCounterStyle.

It is bit unclear to me what CounterStyle class represents. Is it final computed counter style that can be pointed to by RenderStyle and directly used by layout/rendering code? Is it an intermediate type that acts as input for style resolution? Or both?